### ISSUE TITLE
com.hazelcast.jet.sql.impl.connector.map.SqlJoinTest.test_innerJoinKeyAndValue

### ISSUE THREAD

**Issue by olukas** [hazelcast/hazelcast#18846]
_master_ (commit e3352af34221c58de14ed09dcde0edc9206098c8)

Failed on Openj9-11: http://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-master-Openj9-11/29/testReport/com.hazelcast.jet.sql.impl.connector.map/SqlJoinTest/test_innerJoinKeyAndValue/

Stacktrace:
```
java.lang.NullPointerException
	at com.hazelcast.jet.impl.execution.ExecutionContext.terminateExecution(ExecutionContext.java:290)
	at com.hazelcast.jet.impl.JobExecutionService.terminateExecution0(JobExecutionService.java:623)
	at com.hazelcast.jet.impl.JobExecutionService.cancelAllExecutions(JobExecutionService.java:236)
	at com.hazelcast.jet.SimpleTestInClusterSupport.supportAfter(SimpleTestInClusterSupport.java:101)
	at jdk.internal.reflect.GeneratedMethodAccessor28.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at com.hazelcast.test.AbstractHazelcastClassRunner$ThreadDumpAwareRunAfters.evaluate(AbstractHazelcastClassRunner.java:409)
	at com.hazelcast.test.DumpBuildInfoOnFailureRule$1.evaluate(DumpBuildInfoOnFailureRule.java:37)
	at com.hazelcast.test.jitter.JitterRule$1.evaluate(JitterRule.java:104)
	at com.hazelcast.test.metrics.MetricsRule$1.evaluate(MetricsRule.java:62)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at com.hazelcast.test.HazelcastSerialClassRunner.runChild(HazelcastSerialClassRunner.java:50)
	at com.hazelcast.test.HazelcastSerialClassRunner.runChild(HazelcastSerialClassRunner.java:29)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at com.hazelcast.test.AbstractHazelcastClassRunner$1.evaluate(AbstractHazelcastClassRunner.java:306)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:299)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:293)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.lang.Thread.run(Thread.java:836)
```

Standard output:
```
CREATE MAPPING o_71dd3bca_37a7_4444_a314_4a08c2428714 TYPE TestBatch OPTIONS ('names'='v', 'types'='INTEGER', 'values'='0
1
2')
09:13:33,949 DEBUG |test_innerJoinKeyAndValue| - [JetSqlBackend] Thread-328 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Before logical opt:
LogicalProject(v=[$0], id=[$2], name=[$3])
  LogicalJoin(condition=[AND(=($0, $1), =($0, $2))], joinType=[inner])
    LogicalTableScan(table=[[hazelcast, public, o_71dd3bca_37a7_4444_a314_4a08c2428714[projects=[0]]]])
    LogicalProject(__key=[$0], id=[$1], name=[$2])
      LogicalTableScan(table=[[hazelcast, partitioned, o_494f8ef4_9859_4686_88d4_ae6d3b9220d1[projects=[0, 1, 2, 3]]]])

09:13:33,968 DEBUG |test_innerJoinKeyAndValue| - [JetSqlBackend] Thread-328 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] After logical opt:
ProjectLogicalRel(v=[$0], id=[$2], name=[$3])
  JoinLogicalRel(condition=[AND(=($0, $1), =($0, $2))], joinType=[inner])
    FullScanLogicalRel(table=[[hazelcast, public, o_71dd3bca_37a7_4444_a314_4a08c2428714[projects=[0]]]])
    FullScanLogicalRel(table=[[hazelcast, partitioned, o_494f8ef4_9859_4686_88d4_ae6d3b9220d1[projects=[0, 1, 2]]]])

09:13:33,970 DEBUG |test_innerJoinKeyAndValue| - [JetSqlBackend] Thread-328 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] After physical opt:
ProjectPhysicalRel(v=[$0], id=[$2], name=[$3])
  JoinNestedLoopPhysicalRel(condition=[AND(=($0, $1), =($0, $2))], joinType=[inner])
    FullScanPhysicalRel(table=[[hazelcast, public, o_71dd3bca_37a7_4444_a314_4a08c2428714[projects=[0]]]])
    FullScanPhysicalRel(table=[[hazelcast, partitioned, o_494f8ef4_9859_4686_88d4_ae6d3b9220d1[projects=[0, 1, 2]]]])

09:13:33,979  INFO || - [JobCoordinationService] hz.clever_meninsky.cached.thread-3 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Starting job 064b-1b98-5648-0001 based on submit request
09:13:33,985  INFO || - [MasterJobContext] hz.clever_meninsky.cached.thread-3 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Didn't find any snapshot to restore for job '064b-1b98-5648-0001', execution 064b-1b98-5649-0001
09:13:33,985  INFO || - [MasterJobContext] hz.clever_meninsky.cached.thread-3 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Start executing job '064b-1b98-5648-0001', execution 064b-1b98-5649-0001, execution graph in DOT format:
digraph DAG {
	"TestBatch[public.o_71dd3bca_37a7_4444_a314_4a08c2428714]" [localParallelism=1];
	"Join(Lookup-IMap[partitioned.o_494f8ef4_9859_4686_88d4_ae6d3b9220d1])" [localParallelism=2];
	"Project" [localParallelism=2];
	"ClientSink" [localParallelism=1];
	"TestBatch[public.o_71dd3bca_37a7_4444_a314_4a08c2428714]" -> "Join(Lookup-IMap[partitioned.o_494f8ef4_9859_4686_88d4_ae6d3b9220d1])" [label="distributed-partitioned", queueSize=1024];
	"Join(Lookup-IMap[partitioned.o_494f8ef4_9859_4686_88d4_ae6d3b9220d1])" -> "Project" [queueSize=1024];
	"Project" -> "ClientSink" [label="distributed to [127.0.0.1]:5701-partitioned", queueSize=1024];
}
HINT: You can use graphviz or http://viz-js.com to visualize the printed graph.
09:13:33,985 DEBUG || - [MasterJobContext] hz.clever_meninsky.cached.thread-3 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Building execution plan for job '064b-1b98-5648-0001', execution 064b-1b98-5649-0001
09:13:33,986 DEBUG || - [MasterJobContext] hz.clever_meninsky.cached.thread-3 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Built execution plans for job '064b-1b98-5648-0001', execution 064b-1b98-5649-0001
09:13:33,987 DEBUG || - [InitExecutionOperation] hz.keen_meninsky.generic-operation.thread-1 - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] Initializing execution plan for job 064b-1b98-5648-0001, execution 064b-1b98-5649-0001 from [127.0.0.1]:5701
09:13:33,987 DEBUG || - [InitExecutionOperation] hz.clever_meninsky.cached.thread-3 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Initializing execution plan for job 064b-1b98-5648-0001, execution 064b-1b98-5649-0001 from [127.0.0.1]:5701
09:13:33,994  INFO || - [JobExecutionService] hz.keen_meninsky.generic-operation.thread-1 - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] Execution plan for jobId=064b-1b98-5648-0001, jobName='064b-1b98-5648-0001', executionId=064b-1b98-5649-0001 initialized
09:13:33,994  INFO || - [JobExecutionService] hz.clever_meninsky.cached.thread-3 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Execution plan for jobId=064b-1b98-5648-0001, jobName='064b-1b98-5648-0001', executionId=064b-1b98-5649-0001 initialized
09:13:33,994 DEBUG || - [MasterJobContext] hz.clever_meninsky.cached.thread-3 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Init of job '064b-1b98-5648-0001', execution 064b-1b98-5649-0001 was successful
09:13:33,995 DEBUG || - [MasterJobContext] hz.clever_meninsky.cached.thread-3 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Executing job '064b-1b98-5648-0001', execution 064b-1b98-5649-0001
09:13:33,995  INFO || - [JobExecutionService] hz.clever_meninsky.cached.thread-3 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Start execution of job '064b-1b98-5648-0001', execution 064b-1b98-5649-0001 from coordinator [127.0.0.1]:5701
09:13:33,995  INFO || - [JobExecutionService] hz.keen_meninsky.generic-operation.thread-0 - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] Start execution of job '064b-1b98-5648-0001', execution 064b-1b98-5649-0001 from coordinator [127.0.0.1]:5701
09:13:33,999  INFO || - [TaskletExecutionService] hz.clever_meninsky.jet.cooperative.thread-0 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Job was cancelled by the user.
09:13:33,999  INFO |test_innerJoinKeyAndValue| - [SqlTestSupport] Time-limited test - Removing 1 cached plans in SqlTestSupport.@After
09:13:33,999  INFO |test_innerJoinKeyAndValue| - [SqlTestSupport] Time-limited test - Removing 0 cached plans in SqlTestSupport.@After
09:13:33,999 DEBUG || - [JobExecutionService] ForkJoinPool.commonPool-worker-5 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Completed execution of job '064b-1b98-5648-0001', execution 064b-1b98-5649-0001
09:13:34,000  INFO |test_innerJoinKeyAndValue| - [SimpleTestInClusterSupport] Time-limited test - Ditching 4 jobs in SimpleTestInClusterSupport.@After: [064b-1b98-5642-0001, 064b-1b98-5646-0001, 064b-1b98-5644-0001, 064b-1b98-5648-0001]
09:13:34,001 DEBUG |test_innerJoinKeyAndValue| - [Job] Time-limited test - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Sending CANCEL_FORCEFUL request for job 064b-1b98-5648-0001 (name ??)
09:13:33,999 DEBUG || - [JobExecutionService] ForkJoinPool.commonPool-worker-5 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Execution of job '064b-1b98-5648-0001', execution 064b-1b98-5649-0001 completed with failure
java.util.concurrent.CompletionException: java.util.concurrent.CancellationException
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:331) ~[?:?]
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:346) ~[?:?]
	at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:870) ~[?:?]
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837) ~[?:?]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506) ~[?:?]
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2088) ~[?:?]
	at com.hazelcast.jet.impl.util.NonCompletableFuture.internalCompleteExceptionally(NonCompletableFuture.java:72) ~[hazelcast-5.0-SNAPSHOT.jar:5.0-SNAPSHOT]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$ExecutionTracker.taskletDone(TaskletExecutionService.java:486) ~[hazelcast-5.0-SNAPSHOT.jar:5.0-SNAPSHOT]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.dismissTasklet(TaskletExecutionService.java:417) ~[hazelcast-5.0-SNAPSHOT.jar:5.0-SNAPSHOT]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.runTasklet(TaskletExecutionService.java:403) ~[hazelcast-5.0-SNAPSHOT.jar:5.0-SNAPSHOT]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker$$Lambda$1313/0x00000000b45fda70.accept(Unknown Source) ~[?:?]
	at java.util.concurrent.CopyOnWriteArrayList.forEach(CopyOnWriteArrayList.java:807) ~[?:?]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.run(TaskletExecutionService.java:353) ~[hazelcast-5.0-SNAPSHOT.jar:5.0-SNAPSHOT]
	at java.lang.Thread.run(Thread.java:836) ~[?:?]
Caused by: java.util.concurrent.CancellationException
	at com.hazelcast.jet.sql.impl.processors.RootResultConsumerSink.tryProcess(RootResultConsumerSink.java:83) ~[classes/:?]
	at com.hazelcast.jet.impl.execution.ProcessorTasklet.stateMachineStep(ProcessorTasklet.java:295) ~[hazelcast-5.0-SNAPSHOT.jar:5.0-SNAPSHOT]
	at com.hazelcast.jet.impl.execution.ProcessorTasklet.stateMachineStep(ProcessorTasklet.java:280) ~[hazelcast-5.0-SNAPSHOT.jar:5.0-SNAPSHOT]
	at com.hazelcast.jet.impl.execution.ProcessorTasklet.call(ProcessorTasklet.java:254) ~[hazelcast-5.0-SNAPSHOT.jar:5.0-SNAPSHOT]
	at com.hazelcast.jet.impl.execution.TaskletExecutionService$CooperativeWorker.runTasklet(TaskletExecutionService.java:385) ~[hazelcast-5.0-SNAPSHOT.jar:5.0-SNAPSHOT]
	... 4 more
09:13:34,002 DEBUG || - [MasterJobContext] ForkJoinPool.commonPool-worker-5 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] job '064b-1b98-5648-0001', execution 064b-1b98-5649-0001 received response to StartExecutionOperation from [127.0.0.1]:5701: java.util.concurrent.CancellationException
09:13:34,002 DEBUG || - [JobExecutionService] ForkJoinPool.commonPool-worker-5 - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] Completed execution of job '064b-1b98-5648-0001', execution 064b-1b98-5649-0001
09:13:34,002 DEBUG || - [JobExecutionService] ForkJoinPool.commonPool-worker-5 - [127.0.0.1]:5702 [dev] [5.0-SNAPSHOT] Execution of job '064b-1b98-5648-0001', execution 064b-1b98-5649-0001 completed
09:13:34,002 DEBUG || - [MasterJobContext] ForkJoinPool.commonPool-worker-5 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] job '064b-1b98-5648-0001', execution 064b-1b98-5649-0001 received response to StartExecutionOperation from [127.0.0.1]:5702: null
09:13:34,002 DEBUG || - [MasterJobContext] ForkJoinPool.commonPool-worker-5 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] job '064b-1b98-5648-0001', execution 064b-1b98-5649-0001 to be cancelled after Execution
09:13:34,003  INFO || - [MasterJobContext] hz.clever_meninsky.cached.thread-5 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Execution of job '064b-1b98-5648-0001', execution 064b-1b98-5649-0001 got terminated, reason=java.util.concurrent.CancellationException
	Start time: 2021-06-05T09:13:33.979
	Duration: 00:00:00.023
	To see additional job metrics enable JobConfig.storeMetricsAfterJobCompletion
09:13:34,004 DEBUG || - [JobCoordinationService] hz.clever_meninsky.cached.thread-5 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] job '064b-1b98-5648-0001', execution 064b-1b98-5649-0001 is completed
09:13:34,005 DEBUG || - [JobExecutionService] hz.clever_meninsky.async.thread-1 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] job '064b-1b98-5649-0001', execution 064b-1b98-5649-0001 calling completeExecution because execution terminated before it started
09:13:34,005 DEBUG |test_innerJoinKeyAndValue| - [JobExecutionService] Time-limited test - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Completing job '064b-1b98-5649-0001', execution 064b-1b98-5649-0001 locally. Reason: ditching all jobs after a test
09:13:34,005 DEBUG || - [JobExecutionService] hz.clever_meninsky.async.thread-1 - [127.0.0.1]:5701 [dev] [5.0-SNAPSHOT] Completed execution of job '064b-1b98-5649-0001', execution 064b-1b98-5649-0001
BuildInfo right after test_innerJoinKeyAndValue(com.hazelcast.jet.sql.impl.connector.map.SqlJoinTest): BuildInfo{version='5.0-SNAPSHOT', build='20210605', buildNumber=20210605, revision=e3352af, enterprise=false, serializationVersion=1, jet=JetBuildInfo{version='5.0-SNAPSHOT', build='20210605', revision='e3352af'}}
Hiccups measured while running test 'test_innerJoinKeyAndValue(com.hazelcast.jet.sql.impl.connector.map.SqlJoinTest):'
09:13:30, accumulated pauses: 58 ms, max pause: 29 ms, pauses over 1000 ms: 0
```

**Comment by ivanthescientist** [comment]
I tried this on Linux (5.8 kernel) with OpenJDK 8 and bundled IDEA's JDK 11, in both cases the whole suit run and individual test runs succeeded.

**Comment by viliam-durina** [comment]
This issue was fixed by #18825 by the addition of `UNINITIALIZED_LIGHT_JOB_MARKER`, which prevents attempts to cancel uninitialized executions.
